bootstrap/src/basis/file.mli

```diff
@@ -0,0 +1,156 @@
+(** [read_hlt n t] reads up to [n] bytes from [t] and returns the read bytes or
+    halts if bytes could not be read. *)
+
+(* val write: Buffer.t -> ?t !-> Error.t option * ?t *)
+val write: Buffer.t -> t -> Error.t Option.t * t
+(** [write bytes t] writes [bytes] to [t] and returns an error if bytes could
+    not be written. *)
+
+(* val write_hlt: Buffer.t -> ?t !-> ?t *)
+val write_hlt: Buffer.t -> t -> t
+(** [write_htl bytes t] writes [bytes] to [t] and halts if bytes could not be
+    written. *)
+
+(* val close: t !-> Error.t option *)
+val close: t -> Error.t Option.t
+(** [close t] closes file [t] and returns an error if it could not be closed.
+    *)
+
+(* val close: t !-> unit *)
+val close_hlt: t -> unit
+(** [close_hlt t] closes file [t] and halts if it could not be closed. *)
+
+(* val seek: isize -> ?t !-> (usize, Error.t) result * ?t *)
+val seek: ?i:isize -> t -> (usize, Error.t) result * t
+(** [seek n t] mutates file descriptor in [t] to point to the [i]th byte
+    relative to the current byte position of the file. Returns the new byte
+    index relatve to the beginning of the file or an error if it could not be
+    changed. *)
+
+(* val seek_hlt: isize -> ?t !-> ?t *)
+val seek_hlt: ?i:isize -> t -> t
+(** [seek_hlt n t] mutates file descriptor in [t] to point to the [i]th byte
+    relative to the current byte position of the file. Returns the new byte
+    index relatve to the beginning of the file or halts if it could not be
+    changed. *)
+
+(* val seek_hd: isize -> ?t !-> (usize, Error.t) result * ?t *)
+val seek_hd: ?i:isize -> t -> (usize, Error.t) result * t
+(** [seek_left n t] mutates file descriptor in [t] to point to the [u]th byte
+    after to the beginning of the file. Returns the new byte index relatve to
+    the beginning of the file or an error if it could not be changed. *)
+
+(* val seek_hd_hlt: isize -> ?t !-> ?t *)
+val seek_hd_hlt: ?i:isize -> t -> t
+(** [seek_left_hlt n t] mutates file descriptor in [t] to point to the [u]th
+    byte after to the beginning of the file. Returns the new byte index relatve
+    to the beginning of the file or an error if it could not be changed. *)
+
+(* val seek_tl: isize -> ?t !-> (usize, Error.t) result * ?t *)
+val seek_tl: ?i:isize -> t -> (usize, Error.t) result * t
+(** [seek_right n t] mutates file descriptor in [t] to point to the [u]th byte
+    before to the end of the file. Returns the new byte index relatve to the
+    beginning of the file or an error if it could not be changed. *)
+
+(* val seek_tl_hlt: isize -> ?t !-> ?t *)
+val seek_tl_hlt: ?i:isize -> t -> t
+(** [seek_right_hlt n t] mutates file descriptor in [t] to point to the [u]th
+    byte before to the end of the file. Returns the new byte index relatve
+    to the beginning of the file or an error if it could not be changed. *)
+
+module Stream : sig
+  type file = t
+  type t = Buffer.t Stream.t
+
+  (* val read: ?file !-> file * (Buffer.t, Error.t) result Stream.t *)
```

Comment:
```suggestion
  (* val of_file: file -> (Buffer.t, Error.t) result Stream.t *)
```
I'm uneasy about requiring isolated file references here (?file), because it requires some care to keep a reference isolated until it is consumed. Additionally, there's no benefit here, since we return a reference to the file.

Reply:
My intent here was to consume a unique reference to a file. The stream would own it. The `file` that we return here would only be useful for the `File.close` function, basically the interruptable type we mentioned before. Every other function in the API that takes a file input except `File.close` requires a the file to be a unique reference.

Reply:
Unique references are rather precious, and they require meticulous care.  I've been calling them "isolated" rather than "unique", because the only thing that keeps them unique is isolation.  We should definitely chat about this in more detail and figure out whether they're appropriate for this use case.